src/audio/dai-legacy.c

```diff
@@ -705,9 +707,21 @@ static int dai_reset(struct comp_dev *dev)
 {
 	struct dai_data *dd = comp_get_drvdata(dev);
 	struct dma_sg_config *config = &dd->config;
+	int ret;
 
 	comp_info(dev, "dai_reset()");
 
+	/*
+	 * DMA is still running/paused but we have to reset the DAI and release
+	 * the buffer now.
+	 */
+	if (dev->state == COMP_STATE_ACTIVE || dev->state == COMP_STATE_PAUSED) {
+		comp_warn(dev, "dai_reset(): Component is in active state.");
+		ret = dai_comp_trigger(dev, COMP_TRIGGER_STOP);
+		if (ret)
+			return 0;
+	}
+
```

Comment:
While this looks correct, it does seem the root problem is in mux code. Maybe this is an ok safety measure.

This recent kernel PR could help to get the exception dump (if problem happens in rt-resume):
https://github.com/thesofproject/linux/pull/4302 

Btw, this needs to be backported to stable-v2.2 to fix TGL -- we no longer support TGL in SOF mainline with IPC3.

Reply:
It doesn't seem correct to me, there should be something else in the system preventing the system from resetting the DAI while it's active.

My guess is that there are missing patches on the kernel side, or the firmware image is not aligned with what the firmware sends.

Reply:
I see the point now @brentlu but I am still on the side of "let's fix this the right way". This seems to be a fundamental problem that needs to be solved, instead of adding a recovery mechanism for something that should never happen.
This is broken-by-design indeed.

@juimonen @ranj063 FYI.

Reply:
cavs2.5 and other adl-004 or rpl-001 branches.

Reply:
Agree with @plbossart and @kv2019i - this seems like a problem in mux where it also should be fixed. What is the exact sequence of actions here and what would the logical expected behaviour be? Is it the capture (feedback) SSP -> demux -> PCM0 stream that's stopped first and the stop fails because the smart_amp stream is still running? What would we expect here? Does smart amp really absolutely need that feedback stream and it must be kept running or can it be stopped and the smart amp will just continue running in a no-feedback mode?
In the former case we'd expect the feedback stream to be kept alive until the playback stream is freed too, at which point we can free both.
In the latter case we'd want to ignore the feedback link and free the feedback stream and just let the playback pipeline continue? In this case it would seem to me, that we need two kinds of connections: blocking and non-blocking: blocking connections keep the pipeline running if such a blocking connection exists, non-blocking connections can be freed / reset.

Reply:
"Is it the capture (feedback) SSP -> demux -> PCM0 stream that's stopped first and the stop fails because the smart_amp stream is still running? "

Yes, this is what I observed from the log.